### PR TITLE
Allow customizing the static hosts configuration

### PR DESCRIPTION
**Pull request by fantix**
This allows custom runtimes to provide the hosts configuration in their
own way, like using io_uring instead of the `std::fs::File::open()` to
read the system hosts file.

Also publicized the `resolver::system_conf::unix::parse_resolv_conf` function for the same reason.

### CODE REVIEW

File: crates/resolver/src/hosts.rs [hunk 1]
```diff
@@ -109,69 +109,73 @@ fn hosts_path() -> std::path::PathBuf {
...
+        // also empty line also will be ignored,
+        // if line only include `addr` without `host` will be ignored,
+        // the src will be parsed to map in the form `Name -> LookUp`.
+
+        for line in BufReader::new(src).lines() {
+            // Remove comments from the line
+            let line = line
+                .as_ref()
+                .map(|line| line.split('#').next().unwrap().trim())
+                .unwrap_or_default();
```
Comment (bluejekyll) [inline crates/resolver/src/hosts.rs:135]: Let's return this error.

**Review by bluejekyll — CHANGES_REQUESTED** [review]
A couple of minor things. Overall this looks good. Thanks for putting up the PR.

File: crates/resolver/src/hosts.rs [hunk 2]
```diff
@@ -109,69 +109,73 @@ fn hosts_path() -> std::path::PathBuf {
...
-    for line in BufReader::new(file).lines() {
-        // Remove comments from the line
-        let line = line
-            .as_ref()
-            .map(|line| line.split('#').next().unwrap().trim())
-            .unwrap_or_default();
-        if line.is_empty() {
-            continue;
-        }
+impl Hosts {
```
[the remark this answers is not in the document]
Comment (fantix) [reply crates/resolver/src/hosts.rs:117]: Ah, I was trying to keep the git history/diff, but looks like the indent is anyway breaking it, so I'll move it up then.

File: crates/resolver/src/hosts.rs [hunk 3]
```diff
@@ -109,69 +109,73 @@ fn hosts_path() -> std::path::PathBuf {
...
+    /// parse configuration from `src`
+    pub fn read_hosts_conf(mut self, src: impl io::Read) -> Self {
```
[the remark this answers is not in the document]
Comment (fantix) [reply crates/resolver/src/hosts.rs]: Makes sense, fixed for this case!

**Review by bluejekyll — APPROVED** [review]
Looks good. Sorry for the delay.

**Comment by fantix** [comment]
Totally understandable, thanks a lot for merging! It’s pretty exciting to see trust-dns working under io_uring through Cython in my [toy project](https://gitee.com/fantix/kloop) for asyncio, it’s only possible because you did a great job with trust-dns!

**Comment by bluejekyll** [comment]
I've had a lot of help from people like you. Thank you!

**Pull request merged**